src/IO/S3/copyS3File.h

```diff
@@ -60,7 +57,24 @@ void copyDataToS3File(
     const String & dest_key,
     const S3Settings::RequestSettings & settings,
     const std::optional<std::map<String, String>> & object_metadata = std::nullopt,
-    ThreadPoolCallbackRunner<void> schedule_ = {},
+    ThreadPoolCallbackRunner<void> schedule = {},
+    bool for_disk_s3 = false);
+
+/// Try to do native copy (copyS3FileNative()) and fallback to copy through
+/// buffers (copyDataToS3File()) if native copy is not possible and could be
+/// falled back.
+void copyS3FileNativeWithFallback(
```

Comment:
Why can't we just add `ReadSettings` to the list of parameters of `copyS3File()` so `copyS3File()` could do the throttling if it's copying through buffers?

Reply:
OK, let's try to go this way. The patch had been updated.